### ISSUE TITLE
cli_wallet doesn't sign transactions if you only have owner key

### ISSUE THREAD

**Issue by vikramrajkumar** [bitshares/bitshares-core#145]
_From @theoreticalbts on January 7, 2016 21:56_

This applies specifically to `prepare_builder_transaction`, but the problem is in `sign_transaction` so it probably affects just about everything.

_Copied from original issue: cryptonomex/graphene#506_

**Comment by oxarbitrage** [comment]
Here is how to do this only with the owner private key imported into wallet. From a new wallet import your owner private key:

`unlocked >>> import_key oxarbitrage3 XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`

Now you have an account in wallet, you can see its data by:

`get_account oxarbitrage3`

As you don't have active key a transfer will fail by authority:

```
unlocked >>>  transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
 transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
2447694ms th_a       wallet.cpp:1923               sign_transaction     ] Caught exception while broadcasting tx e5cdaaa351c251ca2d11c334571c3cba95da5629:  0 exc
eption: unspecified
3030001 tx_missing_active_auth: missing required active authority
Missing Active Authority 1.2.167006
…
```

Please note that if you use memo and as you don't have memo key either a transfer will fail as:

```
unlocked >>> transfer oxarbitrage3 oxarbitrage.a699 5 BTS "SOME MEMOOO" false
transfer oxarbitrage3 oxarbitrage.a699 5 BTS "have some CORE" false
10 assert_exception: Assert Exception
it != _keys.end(): 
    {}
    th_a  wallet.cpp:665 get_private_key

    {"from":"oxarbitrage3","to":"oxarbitrage.a699","amount":"5","asset_symbol":"BTS","memo":"have some CORE","broadcast":false}
    th_a  wallet.cpp:2029 transfer
unlocked >>> 
```

What you need to do is to send without memo and add your owner public key as an authority to your account. This can be done by the new function that deals with authority:

https://github.com/bitshares/bitshares-core/pull/270

Simply add your public owner key as active as follows:

`update_all_auth oxarbitrage3 [] [] [] [["BTS75uzxS57o2Hey9eJHZ3zXBW9dGDjujPJzcxGaZZqnEUjNmbMct", 1]] 1 1 true`


I am not sure why the results are not reflected immediately in the cli:

`get_account oxarbitrage3`

Will not show the new key while http://cryptofresh.com/u/oxarbitrage3 will display the authority changes made by the cli  immediately. This is probably a topic for another issue.

So i need to restart my wallet, unlock and execute `get_account` again to see the updated owner key in the active spot as:

```
 ...
  "active": {
    "weight_threshold": 1,
    "account_auths": [],
    "key_auths": [[
        "BTS7UEwDk19nbwasN4p95jF15TfEzV61rqoF58BquFgoVt5rt9jxk",
        1
      ],[
        "BTS75uzxS57o2Hey9eJHZ3zXBW9dGDjujPJzcxGaZZqnEUjNmbMct",
        1
      ]
    ],
    "address_auths": []
  },
…
```

With this change transfer will work:

```
unlocked >>> transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
{
  "ref_block_num": 30186,
  "ref_block_prefix": 236223891,
  "expiration": "2017-05-03T13:47:00",
  "operations": [[
      0,{
        "fee": {
          "amount": 264174,
          "asset_id": "1.3.0"
        },
        "from": "1.2.167006",
        "to": "1.2.152768",
        "amount": {
          "amount": 500000,
          "asset_id": "1.3.0"
        },
        "extensions": []
      }
    ]
  ],
  "extensions": [],
  "signatures": [
    "1f50a59fbcfd4617b79377f16840c24a914af281ed9bf471b8523419f9d48d5a7a5ee5b9115b6fadea76d5e79bf41bf31a4a2af18381445221cfdd0bbe83ea9d33"
  ]
}
unlocked >>>
```

**Comment by TheTaconator** [comment]
@oxarbitrage, in your description, could you highlight that the key "BTS75uzxS57o2Hey9eJHZ3zXBW9dGDjujPJzcxGaZZqnEUjNmbMct" is the new active key and not the old owner key?  And, perhaps elaborate on where it came from?

And, maybe describe where the "BTS7UEwDk19nbwasN4p95jF15TfEzV61rqoF58BquFgoVt5rt9jxk" key came from?

**Comment by oxarbitrage** [comment]
Account was created by openledger. Then i got the owner private key(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX) from the web and imported to the cli.

when i do get_account the owner key(BTS75uzxS57o2Hey9eJHZ3zXBW9dGDjujPJzcxGaZZqnEUjNmbMct) and active key(BTS7UEwDk19nbwasN4p95jF15TfEzV61rqoF58BquFgoVt5rt9jxk) are already there.

the problem at that point is that you own the private key for BTS75uzxS57o2Hey9eJHZ3zXBW9dGDjujPJzcxGaZZqnEUjNmbMct but not for BTS7UEwDk19nbwasN4p95jF15TfEzV61rqoF58BquFgoVt5rt9jxk

another option is to just import the private key for active from the light wallet(account -> permissions -> active permissions -> click on public key -> click show.

**Comment by abitmore** [comment]
IMHO the code looking for required keys in `sign_transaction()` is redundant.

The work flow should be: call `get_potential_signatures` API to fetch all possible keys to sign a transaction, then call `get_required_signatures` API with a set of available keys to find out a sufficient subset, then sign the transaction with the subset of keys.

Several bugs are involved here.
1. `get_potential_signatures` API doesn't return owner keys even if they can be used to sign a transaction which requires active authority only.
2. `get_required_signatures` doesn't return the really "required" keys, if a needed key is already in the signature, it will still be returned (#200)
3. sometimes unnecessary keys are returned, for example, if call the API with both owner key and active key, both will be returned, a transaction signed with both keys will still pass [this check](https://github.com/bitshares/bitshares-core/blob/23b8843d2f4e23280e3a5e1db480297758c8c872/libraries/chain/protocol/transaction.cpp#L292-L296) (fix of this check requires a hard fork and could be very tricky, see [this commit](https://github.com/cryptonomex/graphene/commit/f5564b1272149dfe788ef70c672b4b2de375d7f8) for example).

These bugs are affecting the bitshares-ui GUI and other client apps, so I think it's better to fix these than to fix cli_wallet alone.

We can extend `get_required_signatures` API so it will return a tuple which contains these fields:
* required keys (only if provided in available_keys)
* missed (potential) keys (not in available_keys nor in signatures)
* unnecessary signatures in the transaction

I have some unfinished/untested code in my local repo, along with some other changes mixed. I'll share them when done some cleanups (may require some time).

> I am not sure why the results are not reflected immediately in the cli:
> ...
> So i need to restart my wallet, unlock and execute get_account again to see the updated owner key in the active spot as:

See #151.

**Comment by abitmore** [comment]
> As you don't have active key a transfer will fail by authority:
>```
>unlocked >>>  transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
 >transfer oxarbitrage3 oxarbitrage.a699 5 BTS "" true
>2447694ms th_a       wallet.cpp:1923               sign_transaction     ] Caught exception while  broadcasting tx e5cdaaa351c251ca2d11c334571c3cba95da5629:  0 exception: unspecified
>3030001 tx_missing_active_auth: missing required active authority
>Missing Active Authority 1.2.167006
>…
>```

This behavior is exactly the problem mentioned in OP. If change `true` to `false` we can see the transaction hasn't been signed. In BitShares it's designed that owner key is able to do (almost?) anything that active key can do.

**Comment by oxarbitrage** [comment]
except for moving funds ? isn't active permissions the ones that allow a transfer ?

**Comment by abitmore** [comment]
Should be no exception. Owner permission is "higher" than active permission.

**Comment by bitcard** [comment]
@abitmore how to fix it? I meet this problom using smartcoin from blockpay.

**Comment by abitmore** [comment]
By now, you can add your owner key to active permission (if using wallet model but not account model). If you're "using smartcoin from blockpay", before we fix it in backend, you can ask blockpay team to fix on their side, like what's done in light wallet (there is a "sign as owner" option).

**Comment by abitmore** [comment]
Should be fixed with #321 and #497.
